ERCS/erc-7566.md

```diff
@@ -0,0 +1,193 @@
+---
+eip: 7566
+title: Multiplayer Onchain Game
+description: A solution that enables multiple players to play onchain game.
+author: Rickey (@HelloRickey)
+discussions-to: https://ethereum-magicians.org/t/erc-proposal-multiplayer-onchain-game/16796
+status: Draft
+type: Standards Track
+category: ERC
+created: 2023-11-28
+---
+
+## Abstract
+
+This proposal proposes a multiplayer onchain game (MOG) standard interface based on smart contracts, using room to match and group players, and using message to process actions between players. This allows one smart contract to handle multiple players playing games on the chain, preventing centralized servers from affecting the fairness of the game.
+
+## Motivation   
+
+Common multiplayer games are generally played on centralized servers. Players have no way of knowing whether there are forged data and cheating on the server. The owner of the game server can match players at will, modify scores and levels, and even close and pause the game. If the player's actions all occur on the chain, every message from the chain is proof of the player's instructions and actions, which further ensures the fairness of the game. The Multiplayer Onchain Game framework scales vertically by adding rooms to handle and accommodate multiple players. Write on-chain game logic with custom messages for horizontal expansion, allowing game developers to build multiplayer and fully on-chain games with smart contracts.
+
```

Comment:
The biggest outstanding objection I have is that I don't see which parties would need to cooperate on this standard. If I made a game ReallyAwesomeGame, I'd be free to use whatever architecture I want. No one else needs to interact with it besides ReallyAwesomeGame's frontend and the on-chain contracts.

I wouldn't expect Metamask to support ReallyAwesomeGame directly. I wouldn't really expect Etherscan to display the details of ReallyAwesomeGame either.

EIPs really shine when different parties have to interoperate without coordination, like ERC-20 tokens and wallets, or NFTs and marketplaces. I just don't see who would have to work together for this standard.

---

That's not to say you can't get this merged. It just means you need to expand your motivation with information about the different parties that need to work together.

Reply:
Hey @SamWilsn , thank you very much for your different perspective, I updated some of the advantages of using this standard in motivation. Please check it out.

If you build a multiplayer game that doesn't use any standards and don't want etherscan or wallets to support it, I don't see anything wrong with that, just like some people don't like using erc20 and erc721. When standards are adopted on a large scale, many parties will consider how to integrate them into products. I would like to use some examples to explain how other parties can cooperate on this standard when we use a standard to build a game.

Block explorer
When I enter my wallet address, it will query which games I have participated in, and then query which rooms I have joined based on `getRoomIds`, and then display my game behavior records in each room based on `getMemberId` and `getMessageIds`.This game log may be more readable than the transaction log.I can know the result of every game I play, whether I win or lose.

Wallet
When I call the contract `join`, the wallet will remind me whether I want to join the room with Room ID 32, which has a total of 100 people. When I call `sendMessage`, the wallet will prompt me with the contents of Room Id 32, From xxx, To xxx, Message xxxx. The wallet can even provide the Import Rooms function. I can enter the game contract address and Room Id, and the wallet can determine whether the room exists for me through `hasMember`.This is like importing NFT into my wallet, making it easier for me to play games next time.

Guild, Galxe, and other Proof of Attendance Protocol platforms
They can use the game contract address and `hasMember` to determine whether a player has participated in their game. If so, the player can receive a POAP.

Game statistics platform like Defillama
The game activity level of each chain can be analyzed based on `getRoomCount`, `getMemberCount` and `getMessageIds`.

some other applications:
For voting applications, use `createRoom` to create a voting room and define a message type. After participants `join` the room, they vote through `sendMessage`.
For social applications, use `createRoom` to create a topic room. Users entering the room can post content related to this topic through `sendMessage`.